### PR TITLE
Feature: support qwen and llama4 reducescatter for dp attention padding

### PR DESCRIPTION
**Pull request by Misaka9468**
## Motivation
Similar to #8280  and #8539 , this PR adds support for using reduce-scatter instead of all-reduce after MoE/MLP layers in Qwen2 MoE, Qwen3 MoE, and Llama4 when DP attention uses max padding.

## Modifications
As mentioned in #8539 , the reduce-scatter optimization can be extended to other models that use `LayerCommunicator`. Following this approach, I have modified `qwen2_moe.py`, `qwen3_moe.py`, `llama.py`, and `llama4.py` to support this optimization.

Since `llama4.py` reuses the MLP implementation from `llama.py`, the changes in `llama.py` are necessary to ensure consistency and correctness.

work together with my colleague @ttaohe

## Accuracy Tests

For qwen3_moe

```
python3 -m sglang.launch_server --model-path /wxzhoucs/LLM/Qwen3-235B-A22B-Instruct-2507-FP8/ --trust-remote-code --tp 8 --enable-ep-moe --ep-size 8 --dp 8 --enable-dp-attention --port 8008 --mem-fraction-static=0.9 --cuda-graph-max-bs 136 --disable-radix-cache
python3 benchmark/gsm8k/bench_sglang.py --num-shots 8 --num-questions 1319 --parallel 1319 --port=8008

Accuracy: 0.957
Invalid: 0.000
Latency: 47.080 s
Output throughput: 4092.640 token/s
```

For llama4
```
python3 -m sglang.launch_server --model-path /wxzhoucs/LLM/Llama-4-Scout-17B-16E/ --trust-remote-code --tp 8 --dp 8 --enable-dp-attention --port 8008 --mem-fraction-static=0.9 --cuda-graph-max-bs 136 --disable-radix-cache --attention-backend fa3

python3 benchmark/gsm8k/bench_sglang.py --num-shots 8 --num-questions 1319 --parallel 1319 --port=8008

Accuracy: 0.836
Invalid: 0.001
Latency: 78.078 s
```

## Benchmarking and Profiling

* For qwen3_moe
```
python3 -m sglang.launch_server --model-path /wxzhoucs/LLM/Qwen3-235B-A22B-Instruct-2507-FP8/ --trust-remote-code --tp 8 --enable-ep-moe --ep-size 8 --dp 8 --enable-dp-attention --port 8008 --mem-fraction-static=0.9 --cuda-graph-max-bs 136 --disable-radix-cache

python3 -m sglang.bench_serving --backend sglang --dataset-name random --num-prompt 1024 --random-input 1024 --random-output 1024 --random-range-ratio 1 --max-concurrency 1024 --port 8008
```

```
with PR:

============ Serving Benchmark Result ============
Backend:                                 sglang    
Traffic request rate:                    inf       
Max request concurrency:                 1024      
Successful requests:                     1024      
Benchmark duration (s):                  165.23    
Total input tokens:                      1048576   
Total generated tokens:                  1048576   
Total generated tokens (retokenized):    1047231   
Request throughput (req/s):              6.20      
Input token throughput (tok/s):          6346.27   
Output token throughput (tok/s):         6346.27   
Total token throughput (tok/s):          12692.54  
Concurrency:                             1021.66   
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   164849.83 
Median E2E Latency (ms):                 164837.87 
---------------Time to First Token----------------
Mean TTFT (ms):                          25618.14  
Median TTFT (ms):                        25613.09  
P99 TTFT (ms):                           48805.89  
---------------Inter-Token Latency----------------
Mean ITL (ms):                           136.10    
Median ITL (ms):                         112.36    
P95 ITL (ms):                            132.82    
P99 ITL (ms):                            146.58    
Max ITL (ms):                            45650.78  
==================================================


before:

============ Serving Benchmark Result ============
Backend:                                 sglang    
Traffic request rate:                    inf       
Max request concurrency:                 1024      
Successful requests:                     1024      
Benchmark duration (s):                  184.49    
Total input tokens:                      1048576   
Total generated tokens:                  1048576   
Total generated tokens (retokenized):    968873    
Request throughput (req/s):              5.55      
Input token throughput (tok/s):          5683.54   
Output token throughput (tok/s):         5683.54   
Total token throughput (tok/s):          11367.09  
Concurrency:                             1022.09   
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   184148.63 
Median E2E Latency (ms):                 184143.46 
---------------Time to First Token----------------
Mean TTFT (ms):                          28790.29  
Median TTFT (ms):                        28911.27  
P99 TTFT (ms):                           55409.45  
---------------Inter-Token Latency----------------
Mean ITL (ms):                           151.87    
Median ITL (ms):                         125.18    
P95 ITL (ms):                            148.27    
P99 ITL (ms):                            161.61    
Max ITL (ms):                            52526.60  
==================================================

```

* For qwen2_moe

```
python3 -m sglang.launch_server --model-path /wxzhoucs/LLM/Qwen2-57B-A14B-Instruct/ --trust-remote-code --tp 8 --enable-ep-moe --ep-size 8 --dp 8 --enable-dp-attention --port 8008 --mem-fraction-static=0.9 --cuda-graph-max-bs 136 --disable-radix-cache

...
```

```
with PR:

============ Serving Benchmark Result ============
Backend:                                 sglang    
Traffic request rate:                    inf       
Max request concurrency:                 1024      
Successful requests:                     1024      
Benchmark duration (s):                  83.99     
Total input tokens:                      1048576   
Total generated tokens:                  1048576   
Total generated tokens (retokenized):    1047305   
Request throughput (req/s):              12.19     
Input token throughput (tok/s):          12483.85  
Output token throughput (tok/s):         12483.85  
Total token throughput (tok/s):          24967.70  
Concurrency:                             1020.18   
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   83681.47  
Median E2E Latency (ms):                 83633.03  
---------------Time to First Token----------------
Mean TTFT (ms):                          18478.34  
Median TTFT (ms):                        18227.46  
P99 TTFT (ms):                           35616.97  
---------------Inter-Token Latency----------------
Mean ITL (ms):                           63.74     
Median ITL (ms):                         46.87     
P95 ITL (ms):                            61.02     
P99 ITL (ms):                            74.00     
Max ITL (ms):                            32562.22  
==================================================

before:

============ Serving Benchmark Result ============
Backend:                                 sglang    
Traffic request rate:                    inf       
Max request concurrency:                 1024      
Successful requests:                     1024      
Benchmark duration (s):                  99.02     
Total input tokens:                      1048576   
Total generated tokens:                  1048576   
Total generated tokens (retokenized):    743535    
Request throughput (req/s):              10.34     
Input token throughput (tok/s):          10589.05  
Output token throughput (tok/s):         10589.05  
Total token throughput (tok/s):          21178.11  
Concurrency:                             1021.38   
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   98771.27  
Median E2E Latency (ms):                 98747.82  
---------------Time to First Token----------------
Mean TTFT (ms):                          23099.56  
Median TTFT (ms):                        23106.16  
P99 TTFT (ms):                           43464.96  
---------------Inter-Token Latency----------------
Mean ITL (ms):                           73.97     
Median ITL (ms):                         54.04     
P95 ITL (ms):                            71.42     
P99 ITL (ms):                            82.81     
Max ITL (ms):                            40209.97  
==================================================

```

* For llama4

```
python3 -m sglang.launch_server --model-path /wxzhoucs/LLM/Llama-4-Scout-17B-16E/ --trust-remote-code --tp 8 --dp 8 --enable-dp-attention --port 8008 --mem-fraction-static=0.9 --cuda-graph-max-bs 136 --disable-radix-cache --attention-backend fa3
...

```

```
with PR:

============ Serving Benchmark Result ============
Backend:                                 sglang    
Traffic request rate:                    inf       
Max request concurrency:                 1024      
Successful requests:                     1024      
Benchmark duration (s):                  132.41    
Total input tokens:                      1048576   
Total generated tokens:                  1048576   
Total generated tokens (retokenized):    1039307   
Request throughput (req/s):              7.73      
Input token throughput (tok/s):          7919.08   
Output token throughput (tok/s):         7919.08   
Total token throughput (tok/s):          15838.15  
Concurrency:                             1022.04   
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   132158.37 
Median E2E Latency (ms):                 132137.96 
---------------Time to First Token----------------
Mean TTFT (ms):                          19349.94  
Median TTFT (ms):                        19132.36  
P99 TTFT (ms):                           36886.73  
---------------Inter-Token Latency----------------
Mean ITL (ms):                           110.27    
Median ITL (ms):                         92.79     
P95 ITL (ms):                            112.40    
P99 ITL (ms):                            124.59    
Max ITL (ms):                            33875.30  
==================================================

before:

============ Serving Benchmark Result ============
Backend:                                 sglang    
Traffic request rate:                    inf       
Max request concurrency:                 1024      
Successful requests:                     1024      
Benchmark duration (s):                  139.76    
Total input tokens:                      1048576   
Total generated tokens:                  1048576   
Total generated tokens (retokenized):    1029858   
Request throughput (req/s):              7.33      
Input token throughput (tok/s):          7502.87   
Output token throughput (tok/s):         7502.87   
Total token throughput (tok/s):          15005.74  
Concurrency:                             1021.79   
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   139455.13 
Median E2E Latency (ms):                 139418.72 
---------------Time to First Token----------------
Mean TTFT (ms):                          20918.32  
Median TTFT (ms):                        20948.68  
P99 TTFT (ms):                           39783.72  
---------------Inter-Token Latency----------------
Mean ITL (ms):                           115.87    
Median ITL (ms):                         95.90     
P95 ITL (ms):                            123.94    
P99 ITL (ms):                            144.57    
Max ITL (ms):                            36709.55  
==================================================

```

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

### CODE REVIEW

**Comment by Misaka9468** [comment]
@ch-wan  Could you please take a look? Thanks a lot

File: python/sglang/srt/models/qwen2_moe.py [hunk 1]
```diff
@@ -175,7 +181,10 @@ def __init__(
         self.shared_expert_gate = torch.nn.Linear(config.hidden_size, 1, bias=False)
 
     def forward(
-        self, hidden_states: torch.Tensor, forward_batch: Optional[ForwardBatch] = None
+        self,
+        hidden_states: torch.Tensor,
+        forward_batch: Optional[ForwardBatch] = None,
+        use_reduce_scatter: bool = False,
     ) -> torch.Tensor:
```
[the remark this answers is not in the document]
Comment (Misaka9468) [reply python/sglang/srt/models/qwen2_moe.py:186]: But I followed the deepseek_v2.py.

**Comment by Misaka9468** [comment]
Here is an update on the progress of the #9101 PR.

(1) Regarding the CI test failures: I have added a skip_all_reduce parameter to the RowParallelLinearWithLoRA.forward() function in sglang/srt/layers/lora/layers.py. I also added a conditional check for this parameter before the tensor_model_parallel_all_reduce(). With this commit, I believe all CI tests should now pass.

(2) Following the advice from @ch-wan , I ran the test_hybrid_dp_ep_tp_mtp.py script with the following models: Qwen2-57B-A14B-Instruct, Qwen3-235B-A22B, and Llama-4-Scout-17B-16E. I skipped the tests related to DEEPEP and EAGLE, as my code changes do not affect them. Additionally, I skipped the EP test for llama4 because it is not supported by llama4. The results show that all tests passed successfully.

**Review by ch-wan — APPROVED** [review]
LGTM. Thank you for your contribution.
